serve/mlc_serve/engine/engine_common.py

```diff
@@ -135,6 +137,62 @@ def detokenize_incrementally(
     return delta
 
 
+def logprob_detokenize(
+        tokenizer: TokenizerP,
+        logprob_info: Optional[RawLogprobsInfo],
+) -> Optional[LogprobsContent]:
+    """Detokenize tokens from RawLogprobInfo and convert the latter to LogprobContent"""
+    if logprob_info is None:
+        return None
+
+    top_logprobs: List[TopLogprobs] = []
+    if (
+        logprob_info.top_tokens is not None and
+        logprob_info.top_logprobs is not None
+    ):
+        top_tokens = list(zip(logprob_info.top_tokens, logprob_info.top_logprobs))
+        # dedup duplicates
+        # Todo: Make sure decode can generate different tokens
+        if logprob_info.previous_tokens is None:
+            logprob_info.previous_tokens = []
+        for top_token, top_logprob in top_tokens:
+            detokenized = tokenizer.convert_ids_to_tokens(logprob_info.previous_tokens + [top_token])[-1]
+            LOG.info(f"detokenized: {detokenized}")
+            top_logprobs.append(TopLogprobs(
+                token=detokenized,
+                logprob=float(top_logprob),
+                # TODO(vvchernov): implement bytes based on https://platform.openai.com/docs/api-reference/chat/object
+                bytes=None,
+            ))
+
+    logprobs_content = LogprobsContent(
+        token=tokenizer.decode([logprob_info.current_token]),
+        logprob=logprob_info.current_logprob,
+        # TODO(vvchernov): implement bytes based on https://platform.openai.com/docs/api-reference/chat/object
+        bytes=None,
+        top_logprobs=top_logprobs,
+    )
+
+    return logprobs_content
+
+
+def logprobs_detokenize(
+        tokenizer: TokenizerP,
+        logprob_info: List[Optional[RawLogprobsInfo]],
+) -> Optional[List[Optional[LogprobsContent]]]:
+    if logprob_info is None:
+        return None
+
+    res: List[Optional[LogprobsContent]] = []
+    for info in logprob_info:
+        res.append(logprob_detokenize(tokenizer, info))
+
+    check_all = all([x is None for x in res])
+    if check_all:
+        return None
+    return res
```

Comment:
This might be a personal taste but I want to see new lines here between L190 / L191 and L192 / L193. I think that would be more readable.